Delphi 10.2 + Win64 support, JvInterpreter, new defines

idanmiara opened the pull request:
Hi,

In this pull request:
* Added Delphi 10.2 + Win64 support.
* Added jedi.inc (from JEDI) to be included and replace all the Delphi version definitions in JclGlobal.inc 
* Add Pascal Script support (Jv_Interpreter)
* fix some incorrect IFDEFs (set correct Delphi versions)
* add {$DEFINE JCF_GUI}, {$DEFINE JCF_REG} and a minimal JcfGlobal.inc (for apps without GUI and registry support)

Kind Regards,
Idan

ashumkin commented:
Hi! Thank you for your PR! (you're the very first contributor ever!) ))
I've taken a look and I have some comments on your changes. I'll comment them inplace.

ashumkin commented on Include/JcfGlobal.inc:
```diff
@@ -1,4 +1,6 @@
+{$I jedi.inc}
 
+(*
```
I do not like the idea of commenting out the code. Moreover you've deleted it later in c3d5319da9b6c5fc6209b5c855364ccd841aa6cd. So, I'd propose to squash that deletion into this change

idanmiara replied:
thanks for your comment

idanmiara commented:
Thanks!

On Wed, 10 Oct 2018 at 10:38, Alexey Shumkin <notifications@github.com>
wrote:

> Hi! Thank you for your PR! (you're the very first contributor ever!) ))
> I've taken a look and I have some comments on your changes. I'll comment
> them inplace.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ashumkin/jedicodeformat/pull/1#issuecomment-428469778>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AZIQrdloVmoK-u96HD19rJ3qc0DuIXxUks5ujaPygaJpZM4XUtbu>
> .
>

ashumkin commented on JcfGui/fMain.pas:
```diff
@@ -42,11 +42,10 @@ interface
-  , System.Actions
-  {$ENDIF}
-  ;
+  {.$IFDEF DELPHIXE5_UP}
```
What is the reason to use `{.$IFDEF `? Aren't you sure the version of Delphi version where `System.Actions` was introduced? I'm sure this breaks compilation with Delphi 2007, for example

ashumkin replied:
Oh, later the `{$IFDEF DELPHIXE5_UP}` is used, so, you've just missed to remove dot, I guess?

idanmiara replied:
The reason for that comment IFDEF is that there is a bug in (some?) newer versions of Delphi that even if I wrap these definitions in IFDEFs and the define is correct (i.e. DELPHIXE5_UP for Delphi XE5) then when you save the file it adds these units again to the uses clause (and then it won't compile for duplicated uses). So if I need to compile in older version of Delphi I just set this Define.
Another workaround that I use sometimes is to include empty units (like ImageList, Actions) for older Delphi versions and add them in the global search path.
Do you have another idea?

idanmiara replied:
I'm leaving this as is, I couldn't find another workaround for that bug in the IDE.

ashumkin commented on Utils/JcfSystemUtils.pas:
```diff
@@ -102,7 +102,7 @@ function GetWindowsTempFolder: string;
 
 {$ELSE}
-  {$IFDEF WIN32}
+//  {$IFDEF WIN32}
```
Please, do not comment out. Just remove the unused code

idanmiara replied:
point noted.

ashumkin commented on Test/TestCases/PascalScript.pas:
```diff
@@ -0,0 +1,5 @@
+    var
+ RunResult:Variant;
```
Is this file "incorrectly" formatted deliberately? I'd added at least two files: formatted well and "obfuscated"...

ashumkin replied:
Also, I cannot see any other changes to the tests except adding this file. Did I miss something?

idanmiara replied:
Yes, it is "incorrectly" formatted deliberately.
I could add both outputs.

ashumkin commented on JcfGui/fMain.pas:
```diff
@@ -91,6 +91,7 @@   TfrmMain = class(TForm)
     mOutput: TJvMemo;
     lblLog:  TLabel;
     N2:      TMenuItem;
+    Button1: TButton;
```
I do not like the idea of unnamed controls. Please, name it according its function (Caption is `Clear Log`, so `btnClearLog` would be enough)

idanmiara replied:
point noted

ashumkin commented on Parse/BuildParseTree.pas:
```diff
@@ -2913,6 +2913,9 @@ procedure TBuildParseTree.RecogniseStatement;
   CheckLabelPrefix;
 
   lc := fcTokenList.FirstSolidToken;
+  if lc=nil then
```
`lc=nil` this do not follow the style guide concerning operators surrounded by spaces (here the link http://edn.embarcadero.com/article/10280, but surprisingly I have not found this statement there)
So,  I just state: please, surround operators by spaces: `if lc = nil ...` I believe it is more readable.

idanmiara replied:
ok

ashumkin commented on Parse/BuildParseTree.pas:
```diff
@@ -2933,6 +2936,7 @@ procedure TBuildParseTree.CheckLabelPrefix;
   lbColonSecond: boolean;
 begin
   lc2 := fcTokenList.SolidToken(2);
+  if lc2=nil then exit;
```
the same here: `lc2=nil`...
Also, "one statement for a line": `if statement` - one line, `Exit` is on the next...
```delphi
if lc2 = nil then
  Exit;
```

idanmiara replied:
ok

ashumkin reviewed (commented):
The message "minor fix" tells nothing to me about changes the commit brings in. Please, be more specific here.

ashumkin commented on Include/Minimal/JcfGlobal.inc:
```diff
@@ -0,0 +1,3 @@
+{$I jedi.inc}
+{.$DEFINE JCF_GUI}
```
I do not the the idea of having this file. moreover with the "commented" conditionals

idanmiara replied:
I integrated your formatting code in my app, without GUI (which uses JvInterpreter to run scripts).
I don't want my app to be able to read/write to the registry.

Thus, I tried to stick to the bare minimum of formatting code to be added into my app.
It is a good idea to separate the GUI from the formatting logic, but it was too much work to make a complete separation and the use of these IFDEFs was the fastest and easiest solution for that requirement. 

The commented conditions are here for anyone who wishes to use these features.
It's a well known practice.

ashumkin reviewed (commented):
I suppose, those conditionals are introduced to make a programmer to be able to disable some functionality, but I do not even guess why this is needed. Could you be more specific in a commit message to explain why the one may need to disable some functionality?

idanmiara commented:
OK, I think I've solved everything in this new squashed commit.

ashumkin commented:
Oh, I have not supposed you to squash ALL the changes into a single commit...
I follow the "style" of a many atomic commits. Then if one look into the log he would not be puzzled by the single huge change but in the several and small changes.
Could you roll back to the first proposed version (`git reflog` will help) and appy fixes to it?

Commit:
.gitignore - added. that's the default GitHub gitignore for Delphi.

Commit:
Added jedi.inc (from JEDI) to be included and replace all the Delphi version definitions in JclGlobal.inc
add JCF_GUI JCF_REG to GcfGlobal.inc
add {$DEFINE JCF_GUI}, {$DEFINE JCF_REG} and a minimal JcfGlobal.inc (to use with apps without GUI and registry support)
Delphi 10.2 + Win64 support
fix some incorrect IFDEFs (set correct Delphi versions)

JcfGui/fMain.pas - add btnClearLog; weird commented defines explained Compatibility files
JcfSettings - add a function WriteToFile

Commit:
Add Pascal Script support (Jv_Interpreter)
Add Pascal script test and ouputs

idanmiara commented:
done

Commit:
eliminate warnings and hints

idanmiara commented:
Hi,
I think I addressed all the issues that you raised.
Are you still considering?

idanmiara commented:
Hi @ashumkin 
Hope you are well :)
I see that this PR is still open. I think I've addressed your kind comments back then, but if I forgot something let me know.

idanmiara commented:
Hi @ashumkin :)

idanmiara committed:
Create LICENSE.txt
